### PR TITLE
Resolv path for dmcrypt OSDs

### LINKED ISSUE #2320: osd_fragments does not work with dmcrypt devices
ceph-ansible version: v3.1.0rc1
Setup: non-collocated with dmcrypt and no auto discovery

This step in osd_fragments.yml

`
- name: get osd path
  shell: "df | grep {{ item }} | awk '{print $6}' "
  with_items: "{{ devices }}"
  changed_when: false
  failed_when: false
  check_mode: no
  register: osd_path
`

does not work, when the OSDs are encrypted with dmcrypt.
The output of df looks different when the OSDs are encrypted:
`
/dev/sdb5       8.7M   98K  7.9M   2% /var/lib/ceph/osd-lockbox/40727135-93d3-4002-875b-84899a7e32ea
/dev/dm-1       3.7T  110M  3.7T   1% /var/lib/ceph/osd/ceph-6
`
The above task expects a direct mapping like this:
`
/dev/sdb5       8.7M   98K  7.9M   2% /var/lib/ceph/osd/ceph-6
`

### PR DESCRIPTION
**Pull request by czunker**
This fixes #2320

Signed-off-by: Christian Zunker <christian.zunker@codecentric.de>

### CODE REVIEW

File: roles/ceph-osd/tasks/get_osd_paths_dmcrypt.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  - name: get osd dmcrypt device from uuid
+    shell: "readlink -f /dev/mapper/{{ item.stdout }}"
+    register: dmcrypt_devices
+    changed_when: false
+    failed_when: false
+    check_mode: no
+    with_items: "{{ osd_uuids.results }}"
+
+  - name: get osd path for dmcrypt devices
+    shell: "df | grep {{ item.stdout }} | awk '{print $6}'"
```
Comment (leseb) [inline roles/ceph-osd/tasks/get_osd_paths_dmcrypt.yml]: ditto

File: roles/ceph-osd/tasks/get_osd_paths_dmcrypt.yml [hunk 2]
```diff
@@ -0,0 +1,32 @@
+---
+  - name: get lockbox path for osd devices
+    shell: "df | grep {{ item }} | awk '{print $6}'"
```
[the remark this answers is not in the document]
Comment (guits) [reply roles/ceph-osd/tasks/get_osd_paths_dmcrypt.yml]: just an idea, maybe you could consider using `blkid` like this:

```
[root@osd1 ~]# blkid -s PARTUUID -t PARTLABEL="ceph data" -o value /dev/sda1
7304d8bc-de76-4507-97fc-3684bccc8a4b
[root@osd1 ~]#
```

therefore the path you are looking for would be `/var/lib/ceph/osd-lockbox/7304d8bc-de76-4507-97fc-3684bccc8a4b` (I guess this should be adapted if you are deploying filestore.)


an alternative could be :

```
[root@osd1 ~]# lsblk -no MOUNTPOINT /dev/sda5
/var/lib/ceph/osd-lockbox/7304d8bc-de76-4507-97fc-3684bccc8a4b
[root@osd1 ~]#
```

not sure which option is the best to achieve this (by the way `ceph-volume` is going to change that kind of task in the future) but I generally prefer to privilege using anything before `awk` command.

File: roles/ceph-osd/tasks/get_osd_paths_dmcrypt.yml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    check_mode: no
+    with_items: "{{ osd_uuids.results }}"
+
+  - name: get osd path for dmcrypt devices
+    shell: "df | grep {{ item.stdout }} | awk '{print $6}'"
+    with_items: "{{ dmcrypt_devices.results }}"
+    changed_when: false
+    failed_when: false
+    check_mode: no
+    register: osd_path
```
[the remark this answers is not in the document]
Comment (czunker) [reply roles/ceph-osd/tasks/get_osd_paths_dmcrypt.yml]: These two registers are triggered in two different cases. The one in osd_fragment.yml is called when dmcrypt is not used and the one in get_osd_paths_dmcrypt.yml is called when dmcrypt is used.

This way the following step can use the same variable.

File: roles/ceph-osd/tasks/get_osd_paths_dmcrypt.yml (hunk 1 above)
Comment (czunker) [reply roles/ceph-osd/tasks/get_osd_paths_dmcrypt.yml]: I can't. `command` does not support pipe.

File: roles/ceph-osd/tasks/get_osd_paths_dmcrypt.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (czunker) [reply roles/ceph-osd/tasks/get_osd_paths_dmcrypt.yml]: @guits Thanks for the hint about lsblk. I will push an update.

File: roles/ceph-osd/tasks/get_osd_paths_dmcrypt.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (guits) [reply roles/ceph-osd/tasks/get_osd_paths_dmcrypt.yml]: @czunker I'm afraid @leseb is right, but I think (unless I missed something) leseb meant to say the second will override the first in the case where it's skipped, `osd_path` will be registered with attributes like `osd_path.skipped = true`

File: roles/ceph-osd/tasks/osd_fragment.yml [hunk 4]
```diff
@@ -1,6 +1,6 @@
 ---
 - name: get osd path
-  shell: "df | grep {{ item }} | awk '{print $6}'"
+  shell: "lsblk -no MOUNTPOINT {{ item }} | grep 'ceph-'"
```
Comment (guits) [inline roles/ceph-osd/tasks/osd_fragment.yml]: @czunker why using `| grep 'ceph-'` while you know the list of devices (`with_items: "{{ devices }}"`) ?

File: roles/ceph-osd/tasks/get_osd_paths_dmcrypt.yml (hunk 1 above)
Comment (leseb) [reply roles/ceph-osd/tasks/get_osd_paths_dmcrypt.yml]: ditto for awk instead of grep+awk

File: roles/ceph-osd/tasks/osd_fragment.yml (hunk 4 above)
Comment (czunker) [reply roles/ceph-osd/tasks/osd_fragment.yml]: @guits With encrypted devices, lsblk returns two entries on my systems:
`
# lsblk -no MOUNTPOINT /dev/sdb
/var/lib/ceph/osd/ceph-6
/var/lib/ceph/osd-lockbox/40727135-93d3-4002-875b-84899a7e32ea
`
I tried to include the partlabel which you used in the blkid example, but lsblk does not show the ceph-data label. So I had to limit the output by grep.

Comment (guits) [reply roles/ceph-osd/tasks/osd_fragment.yml]: @czunker add `--nodeps` option and you will have only 1 entry

Comment (czunker) [reply roles/ceph-osd/tasks/osd_fragment.yml]: @guits It's not working:
```
# lsblk -no MOUNTPOINT --nodeps /dev/sdb

```
This is the complete output for one disk:
```
sdb                                        8:16   0   3.7T  0 disk  
├─sdb1                                     8:17   0   3.7T  0 part  
│ └─40727135-93d3-4002-875b-84899a7e32ea 252:1    0   3.7T  0 crypt /var/lib/ceph/osd/ceph-6
└─sdb5                                     8:21   0    10M  0 part  /var/lib/ceph/osd-lockbox/40727135-93d3-4002-875b-84899a7e32ea
```
I found a way to show the `ceph data` label, but that's assigned to the unencrypted partition:
```
sdb                                        3.7T  0 disk               
├─sdb1                                     3.7T  0 part  ceph data    
│ └─40727135-93d3-4002-875b-84899a7e32ea   3.7T  0 crypt              /var/lib/ceph/osd/ceph-6
└─sdb5                                      10M  0 part  ceph lockbox /var/lib/ceph/osd-lockbox/40727135-93d3-4002-875b-84899a7e32ea
```

Comment (D3N14L) [reply roles/ceph-osd/tasks/osd_fragment.yml]: @guits @czunker   `--nodeps` gives just one entry , but only if I address the correct partition..., but I guess thats not what is in `{{ devices }}` right?
```

## Without nodeps
# lsblk -no MOUNTPOINT /dev/sdb


/var/lib/ceph/osd/ceph-9
/var/lib/ceph/osd-lockbox/ab71d7dc-2b98-4ef4-90cb-23204d7eee50
# lsblk -no MOUNTPOINT /dev/sdb1

/var/lib/ceph/osd/ceph-9
# lsblk -no MOUNTPOINT /dev/sdb5
/var/lib/ceph/osd-lockbox/ab71d7dc-2b98-4ef4-90cb-23204d7eee50




## With nodeps
# lsblk -no MOUNTPOINT --nodeps /dev/sdb

# lsblk -no MOUNTPOINT --nodeps /dev/sdb1

# lsblk -no MOUNTPOINT --nodeps /dev/sdb5
/var/lib/ceph/osd-lockbox/ab71d7dc-2b98-4ef4-90cb-23204d7eee50
```

Comment (czunker) [reply roles/ceph-osd/tasks/osd_fragment.yml]: @D3N14L Yes, devices only includes /dev/sdb and alike.

**Comment by leseb** [comment]
This will be deprecated once we merge https://github.com/ceph/ceph-ansible/pull/2409, @czunker do you still want to see this? Any comments on https://github.com/ceph/ceph-ansible/pull/2409? Thanks

**Comment by czunker** [comment]
@leseb No, thanks. No need to fix some deprecated functionality.

At first sight, no comments to #2409 We will try it with one of the next releases including the PR.

**Comment by leseb** [comment]
@czunker thanks!

**Pull request closed without merging**